fairscale/experimental/nn/auto_shard.py

```diff
@@ -0,0 +1,186 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+from functools import reduce
+import logging
+import operator
+from typing import Any, Dict, List
+
+import torch
+import torch.fx
+from torch.fx.node import Node
+
+
+def _get_count(param_count: Dict, node_name: Any) -> int:
+    """Identify different mutations of a given node name."""
+    if node_name in param_count:
+        return param_count[node_name]
+    elif node_name.split("_")[0] in param_count:
```

Comment:
When modules are not explicitly named, the traced graph prepends an "_" to the layer name. I am not happy with this code since I think it may lead to errors. Working on finding a better way to track nodes in the graph. I added a TODO to track this.

Reply:
A valid name could also have an `_` in it. There could multiple names with a common prefix followed by an `_`.

The `_` handling seems unnecessary. Why is it needed? In the fx examples, a dictionary is created directly from `named_modules`:

```
self.modules = dict(self.mod.named_modules())
```
Reference: https://pytorch.org/docs/stable/fx.html#the-interpreter-pattern

Reply:
The handling is necessary because we need a way to connect the node we are tracing with the node specified in the param count dictionary. When you symbolically trace a graph, the nodes names assigned need to match the ones from named_modules(). If i can get the param count when I trace the graph that would mean a SSOT for the name.

Reply:
The[ code](https://pytorch.org/docs/stable/fx.html#the-interpreter-pattern) linked to above suggests that the names in the graph do match the names from `named_modules()`. Are you seeing something different?

Reply:
Thats right. I am seeing something different that needs more investigation and discussion with the torch.fx team.